### PR TITLE
[core] Include SARIF renderer

### LINKED ISSUE #2953: [core] Support SARIF JSON Format
**Is your feature request related to a problem? Please describe.**
SARIF is an OASIS Standard for static analysis alerts. GitHub support uploading SARIF formatted alerts to their CodeScanning API which allows alerts to be surfaced in the Github UI.

**Describe the solution you'd like**
Enable a flag for outputting reports in SARIF.

**Describe alternatives you've considered**

**Additional context**
https://docs.github.com/en/free-pro-team@latest/github/finding-security-vulnerabilities-and-errors-in-your-code/uploading-a-sarif-file-to-github
https://sarifweb.azurewebsites.net/Validation

### PR DESCRIPTION
**Pull request by mmoyaferrer**
## Support SARIF Output Format 

Inclusion of a SARIF renderer in pmd-core

SARIF 2.1.0 Schema:
- https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json

SARIF Docs:

- https://github.com/microsoft/sarif-tutorials/tree/main/docs

SARIF Validators:
- https://sarifweb.azurewebsites.net/Validation
- https://microsoft.github.io/sarif-web-component/
- https://github.com/microsoft/sarif-vscode-extension

## Related issues

- Fixes #2953

## Ready?

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by travis)
- [x] Added (in-code) documentation (if needed)

### CODE REVIEW

**Comment by rsoesemann** [comment]
@josepalafox & @swinton:

@mmoyaferrer and me would also love to get your feedback on our v1 implementation of SARIF.

**Comment by mmoyaferrer** [comment]
Example of PMD Sarif Output (generated with snapshot bin):

```
{
  "$schema": "https://raw.githubusercontent.com/oasis-tcs/sarif-spec/master/Schemata/sarif-schema-2.1.0.json",
  "version": "2.1.0",
  "runs": [
    {
      "tool": {
        "driver": {
          "name": "PMD",
          "version": "6.31.0-SNAPSHOT",
          "informationUri": "https://pmd.github.io/pmd/",
          "rules": [
            {
              "id": "ApexSharingViolations",
              "shortDescription": {
                "text": "Apex classes should declare a sharing model if DML or SOQL/SOSL is used"
              },
              "helpUri": "https://pmd.github.io/pmd/pmd_rules_apex_security.html#apexsharingviolations",
              "properties": {
                "ruleset": "Security",
                "priority": 3
              }
            },
            {
              "id": "ApexDoc",
              "shortDescription": {
                "text": "Missing ApexDoc comment"
              },
              "helpUri": "https://pmd.github.io/pmd/pmd_rules_apex_documentation.html#apexdoc",
              "properties": {
                "ruleset": "Documentation",
                "priority": 3
              }
            }
          ]
        }
      },
      "results": [
        {
          "ruleId": "ApexSharingViolations",
          "ruleIndex": 0,
          "message": {
            "text": "Apex classes should declare a sharing model if DML or SOQL/SOSL is used"
          },
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "/Users/manuelmoyaferrer/devel/pmd-copado/unhappy-soup/force-app/main/default/classes/ApexSOQLInjection.cls"
                },
                "region": {
                  "startLine": 1,
                  "startColumn": 14,
                  "endLine": 1,
                  "endColumn": 30
                }
              }
            },
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "/Users/manuelmoyaferrer/devel/pmd-copado/unhappy-soup/force-app/main/default/classes/ApexSharingViolations.cls"
                },
                "region": {
                  "startLine": 1,
                  "startColumn": 14,
                  "endLine": 1,
                  "endColumn": 34
                }
              }
            }
          ]
        },
        {
          "ruleId": "ApexDoc",
          "ruleIndex": 1,
          "message": {
            "text": "Missing ApexDoc comment"
          },
          "locations": [
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "/Users/manuelmoyaferrer/devel/pmd-copado/unhappy-soup/force-app/main/default/classes/ApexSOQLInjection.cls"
                },
                "region": {
                  "startLine": 1,
                  "startColumn": 14,
                  "endLine": 6,
                  "endColumn": 1
                }
              }
            },
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "/Users/manuelmoyaferrer/devel/pmd-copado/unhappy-soup/force-app/main/default/classes/ApexSOQLInjection.cls"
                },
                "region": {
                  "startLine": 3,
                  "startColumn": 17,
                  "endLine": 5,
                  "endColumn": 2
                }
              }
            },
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "/Users/manuelmoyaferrer/devel/pmd-copado/unhappy-soup/force-app/main/default/classes/ApexSharingViolations.cls"
                },
                "region": {
                  "startLine": 1,
                  "startColumn": 14,
                  "endLine": 6,
                  "endColumn": 1
                }
              }
            },
            {
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "/Users/manuelmoyaferrer/devel/pmd-copado/unhappy-soup/force-app/main/default/classes/ApexSharingViolations.cls"
                },
                "region": {
                  "startLine": 3,
                  "startColumn": 14,
                  "endLine": 5,
                  "endColumn": 2
                }
              }
            }
          ]
        }
      ]
    }
  ]
}
```

**Comment by rsoesemann** [comment]
@mmoyaferrer can you solve the unsuccessful check that are reported above?

File: pmd-core/pom.xml [hunk 1]
```diff
@@ -190,6 +195,12 @@
             <artifactId>system-rules</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-annotations</artifactId>
+            <version>2.6.0</version>
+            <scope>compile</scope>
+        </dependency>
```
Comment (oowekyala) [inline pmd-core/pom.xml:203]:
```suggestion
```

I think this is unused now

**Review by oowekyala — CHANGES_REQUESTED** [review]
Hi, thanks for working on this!

I don't like that you added lombok for this though. It looks like it saves space in the schema, but we don't use lombok anywhere else and I'd like to keep it that way, at least for now. If you want to keep using lombok, then those classes should be in another library. If that's too much of a hassle, then please desugar them, and put them into an internal package. An IDE can do that automatically.

File: pmd-core/pom.xml (hunk 1 above)
Comment (rsoesemann) [reply pmd-core/pom.xml:203]: Thanks @oowekyala for looking into this. We will remove the dependency very quickly. Did you check the rest of the PR? And like it?

Comment (oowekyala) [reply pmd-core/pom.xml:203]: Yes this is nice, it's well tested and documented :)

**Review by adangel — CHANGES_REQUESTED** [review]
Thanks for working on this!

I've three points:

* I'd also prefer not to use lombok. You can either [delombok](https://projectlombok.org/features/delombok) it - or alternatively, we could use a library that brings in the SARIF format, e.g. https://github.com/Contrast-Security-OSS/java-sarif (this actually uses lombok as well, but since we would use this lib as a dependency, that would be ok) or https://github.com/VillageChief/sarif-java . See also https://search.maven.org/search?q=sarif for the available libs.

* Please also remove the ".DS_Store" files/directories. Maybe add this even to the top-level `.gitignore` file.

* The build + tests seems to run successfully under linux, just Windows has problems... I guess, it's about the different line endings. For the JSON renderer, we solved that by implementing the filter-method in the test, which normalizes the line endings:

https://github.com/pmd/pmd/blob/f8c091b7426cbfaa6b3ce79a6e9995bee3daad32/pmd-core/src/test/java/net/sourceforge/pmd/renderers/JsonRendererTest.java#L83

That means, the expected json files are stored with "\n", and the produced output of the renderer in the test will be normalized to the same line ending.

I can confirm, that this fixes the unit tests.

File: pmd-core/pom.xml (hunk 1 above)
Comment (mmoyaferrer) [reply pmd-core/pom.xml]: Nice catch! Removed.

File: pmd-core/pom.xml [hunk 2]
```diff
@@ -147,6 +147,11 @@
             <version>2.8.5</version>
         </dependency>
 
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
```
[the remark this answers is not in the document]
Comment (mmoyaferrer) [reply pmd-core/pom.xml]: Lombok has been removed

File: pmd-core/src/main/java/net/sourceforge/pmd/renderers/SarifRenderer.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+import java.io.IOException;
+import java.util.Iterator;
+
+import net.sourceforge.pmd.Report;
+import net.sourceforge.pmd.RuleViolation;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+
+public class SarifRenderer extends AbstractIncrementingRenderer {
```
[the remark this answers is not in the document]
Comment (mmoyaferrer) [reply pmd-core/src/main/java/net/sourceforge/pmd/renderers/SarifRenderer.java:18]: I totally agree with your point. As a future improvement a new `AbstractIncrementingRenderer` could be created to give this optimization to the SARIF use case (the current `AbstractIncrementingRenderer` does not support it for SARIF)

File: pmd-core/src/main/java/net/sourceforge/pmd/renderers/SarifLog.java [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.Singular;
+import lombok.experimental.Accessors;
+
+@Data
+@Accessors(chain = true)
+@Builder
+public class SarifLog {
```
[the remark this answers is not in the document]
Comment (mmoyaferrer) [reply pmd-core/src/main/java/net/sourceforge/pmd/renderers/SarifLog.java]: Moved, thanks for pointing it out.

File: pmd-core/src/main/java/net/sourceforge/pmd/renderers/SarifRenderer.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    public static final String NAME = "sarif";
+    private static final String DEFAULT_DESCRIPTION = "Sarif integration.";
```
[the remark this answers is not in the document]
Comment (mmoyaferrer) [reply pmd-core/src/main/java/net/sourceforge/pmd/renderers/SarifRenderer.java]: Fixed! :)

**Comment by rsoesemann** [comment]
@adangel @oowekyala Lombok is removed. If you are happy with the rest please merge it. Or should I just do it?
I am unsure about all the housekeeping (documentation, etc...) needed to get this into 6.31.

**Review by adangel — APPROVED** [review]
Thanks, looks good now!

I'll merge it probably on Thursday.

**Review by oowekyala — APPROVED** [review]
I'm glad this will be part of 6.31, thanks for your work

**Pull request merged**